packages/anchor-listener/src/listener.ts

```diff
@@ -0,0 +1,97 @@
+import { ANCHOR_CONTRACT_ADDRESSES } from '@ceramicnetwork/anchor-contract'
+import type { Block, Provider } from '@ethersproject/providers'
+import type { CID } from 'multiformats/cid'
+import {
+  type Observable,
+  type RetryConfig,
+  firstValueFrom,
+  fromEvent,
+  map,
+  mergeMap,
+  scan,
+} from 'rxjs'
+
+import { createRootCIDsLoader, mapLoadBlock } from './loader.js'
+
+export type ProcessedBlockEvent = {
+  type: 'processed-block'
+  block: Block
+  roots: Array<CID>
+}
+
+export type ReorganizationEvent = {
+  type: 'reorganization'
+  latestProcessedBlock: string
+  newBlock: Block
+}
+
+export type ListenerEvent = ProcessedBlockEvent | ReorganizationEvent
+
+export type ListenerParams = {
+  latestProcessedBlock?: string
+  confirmations: number
+  network: keyof typeof ANCHOR_CONTRACT_ADDRESSES
+  provider: Provider
+  retryConfig?: RetryConfig
+}
+
+type State = { status: 'process' | 'reorganized'; previousHash: string; block: Block }
+type Seed = { status: 'initial' }
+
+export function createBlockListener({
```

Comment:
just wanted to mention taking a glance at the implementation. I have read (and seen although it has been a few years) that blocks may come out of order.

Reply:
If so that should trigger the `reorganization` event, now? The logic to detect these is just comparing the newly loaded block's `parentHash` with the hash of the previously processed block.
Shall we change this behavior? If so, what logic should we apply to detect reorganizations please?

Reply:
sorry for the long text, this is me thinking out loud haha

Reply:
Hopefully it's handled as expected now in the listener logic, does it look OK to you please?

Reply:
the logic looks okay to me!

Reply:
Genuine question: does it matter if we process blocks out of order?  We might learn about newer documents or updates before older ones, but does that matter?  We've never really defined what the consistency model across documents is for Ceramic.  I think even if we do the work to keep updates in order here, when we eventually switch away from blockchain anchors as the mechanism for syncing towards a peer-to-peer sync based on looking up providers for a Model in the DHT, then I think we'll struggle to maintain that kind of consistency or ordering guarantee.  It might be good to get our users used to a looser consistency model now.

CC @oed @aarongoldman

Reply:
I have not followed the entire conversation here, but I don't see any reason we couldn't process multiple blocks in parallel.

Reply:
The listener needs to keep track of the block order in order to detect possible reorganizations of the chain, that should be handled by the consumer.
If we processed blocks out of order, we'd need more logic to track previously seen blocks and possible mismatches in the chain ancestry.

Reply:
> If we processed blocks out of order, we'd need more logic to track previously seen blocks and possible mismatches in the chain ancestry.

sorry, I don't really follow - can you elaborate on what this means?

Isn't whether or not we allow blocks to be processed in parallel a separate concern from detecting block reorgs?

Reply:
The listener logic only keeps track of the hash of the previously seen block, to check if `block.parentHash = previousHash` when receiving new blocks to flag reorganizations.
The loading of blocks and anchors proofs can be done in parallel, but then the data gets pushed ordered by block number.

Reply:
okay, that makes sense.  I believe we do still hold off delivering newer blocks until we hear about the older blocks to fill in the gap, but that should be a small amount of time and relatively rare, so probably not a big deal

Reply:
In theory we could process everything except the most recent days worth of blocks in parallel since reorgs are extremely unlikely. However, adding this logic is probably more effort than it's worth right now.